File: qpid/python/tests/broker.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
#   http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
from qpid.client import Closed
from qpid.queue import Empty
from qpid.content import Content
from qpid.testlib import testrunner, TestBase

class BrokerTests(TestBase):
    """Tests for basic Broker functionality"""

    def test_amqp_basic_13(self):
        """
        First, this test tries to receive a message with a no-ack
        consumer. Second, this test tries to explicitely receive and
        acknowledge a message with an acknowledging consumer.
        """
        ch = self.channel
        self.queue_declare(ch, queue = "myqueue")

        # No ack consumer
        ctag = ch.basic_consume(queue = "myqueue", no_ack = True).consumer_tag
        body = "test no-ack"
        ch.basic_publish(routing_key = "myqueue", content = Content(body))
        msg = self.client.queue(ctag).get(timeout = 5)
        self.assert_(msg.content.body == body)

        # Acknowleding consumer
        self.queue_declare(ch, queue = "otherqueue")
        ctag = ch.basic_consume(queue = "otherqueue", no_ack = False).consumer_tag
        body = "test ack"
        ch.basic_publish(routing_key = "otherqueue", content = Content(body))
        msg = self.client.queue(ctag).get(timeout = 5)
        ch.basic_ack(delivery_tag = msg.delivery_tag)
        self.assert_(msg.content.body == body)
        
    def test_basic_delivery_immediate(self):
        """
        Test basic message delivery where consume is issued before publish
        """
        channel = self.channel
        self.exchange_declare(channel, exchange="test-exchange", type="direct")
        self.queue_declare(channel, queue="test-queue") 
        channel.queue_bind(queue="test-queue", exchange="test-exchange", routing_key="key")
        reply = channel.basic_consume(queue="test-queue", no_ack=True)
        queue = self.client.queue(reply.consumer_tag)

        body = "Immediate Delivery"
        channel.basic_publish(exchange="test-exchange", routing_key="key", content=Content(body), immediate=True)
        msg = queue.get(timeout=5)
        self.assert_(msg.content.body == body)

        # TODO: Ensure we fail if immediate=True and there's no consumer.


    def test_basic_delivery_queued(self):
        """
        Test basic message delivery where publish is issued before consume
        (i.e. requires queueing of the message)
        """
        channel = self.channel
        self.exchange_declare(channel, exchange="test-exchange", type="direct")
        self.queue_declare(channel, queue="test-queue")
        channel.queue_bind(queue="test-queue", exchange="test-exchange", routing_key="key")
        body = "Queued Delivery"
        channel.basic_publish(exchange="test-exchange", routing_key="key", content=Content(body))
        reply = channel.basic_consume(queue="test-queue", no_ack=True)
        queue = self.client.queue(reply.consumer_tag)
        msg = queue.get(timeout=5)
        self.assert_(msg.content.body == body)

    def test_invalid_channel(self):
        channel = self.client.channel(200)
        try:
            channel.queue_declare(exclusive=True)
            self.fail("Expected error on queue_declare for invalid channel")
        except Closed, e:
            self.assertConnectionException(504, e.args[0])
        
    def test_closed_channel(self):
        channel = self.client.channel(200)
        channel.channel_open()
        channel.channel_close()
        try:
            channel.queue_declare(exclusive=True)
            self.fail("Expected error on queue_declare for closed channel")
        except Closed, e:
            self.assertConnectionException(504, e.args[0])

File: qpid/cpp/src/qpid/management/Broker.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
 
#include "config.h"
#include "qpid/broker/Broker.h"
#include "qpid/framing/FieldTable.h"
#include "Broker.h"
#include "ArgsBrokerEcho.h"

using namespace qpid::management;
using namespace qpid::sys;
using namespace qpid::framing;

bool Broker::schemaNeeded = true;

Broker::Broker (Manageable* _core, const Options& _conf) :
    ManagementObject (_core, "broker")
{
    broker::Broker::Options& conf = (broker::Broker::Options&) _conf;

    sysId = "sysId";
    port                 = conf.port;
    workerThreads        = conf.workerThreads;
    maxConns             = conf.maxConnections;
    connBacklog          = conf.connectionBacklog;
    stagingThreshold     = conf.stagingThreshold;
    storeLib             = conf.store;
    asyncStore           = conf.storeAsync;
    mgmtPubInterval      = conf.mgmtPubInterval;
    initialDiskPageSize  = 0;
    initialPagesPerQueue = 0;
    clusterName          = "";
    version              = PACKAGE_VERSION;
}

Broker::~Broker () {}

void Broker::writeSchema (Buffer& buf)
{
    FieldTable ft;
    FieldTable arg;

    schemaNeeded = false;

    // Schema class header:
    buf.putShortString (className);  // Class Name
    buf.putShort       (13);         // Config Element Count
    buf.putShort       (0);          // Inst Element Count
    buf.putShort       (1);          // Method Count
    buf.putShort       (0);          // Event Count

    // Config Elements
    ft = FieldTable ();
    ft.setString ("name",   "systemRef");
    ft.setInt    ("type",   TYPE_U64);
    ft.setInt    ("access", ACCESS_RC);
    ft.setInt    ("index",  1);
    ft.setString ("desc",   "System ID");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "port");
    ft.setInt    ("type",   TYPE_U16);
    ft.setInt    ("access", ACCESS_RC);
    ft.setInt    ("index",  1);
    ft.setString ("desc",   "TCP Port for AMQP Service");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "workerThreads");
    ft.setInt    ("type",   TYPE_U16);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Thread pool size");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "maxConns");
    ft.setInt    ("type",   TYPE_U16);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Maximum allowed connections");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "connBacklog");
    ft.setInt    ("type",   TYPE_U16);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Connection backlog limit for listening socket");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "stagingThreshold");
    ft.setInt    ("type",   TYPE_U32);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Broker stages messages over this size to disk");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "storeLib");
    ft.setInt    ("type",   TYPE_SSTR);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Name of persistent storage library");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "asyncStore");
    ft.setInt    ("type",   TYPE_U8);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Use async persistent store");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "mgmtPubInterval");
    ft.setInt    ("type",   TYPE_U16);
    ft.setInt    ("access", ACCESS_RW);
    ft.setInt    ("index",  0);
    ft.setInt    ("min",    1);
    ft.setString ("unit",   "second");
    ft.setString ("desc",   "Interval for management broadcasts");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "initialDiskPageSize");
    ft.setInt    ("type",   TYPE_U32);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Number of disk pages allocated for storage");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "initialPagesPerQueue");
    ft.setInt    ("type",   TYPE_U32);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Number of disk pages allocated per queue");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "clusterName");
    ft.setInt    ("type",   TYPE_SSTR);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Name of cluster this server is a member of, zero-length for standalone server");
    buf.put (ft);

    ft = FieldTable ();
    ft.setString ("name",   "version");
    ft.setInt    ("type",   TYPE_SSTR);
    ft.setInt    ("access", ACCESS_RO);
    ft.setInt    ("index",  0);
    ft.setString ("desc",   "Running software version");
    buf.put (ft);

    // Inst Elements

    return;  // TODO - Remove

    // Methods
    ft = FieldTable ();
    ft.setString ("name",   "echo");
    ft.setInt    ("args",   2);
 
    arg = FieldTable ();
    arg.setString ("name", "sequence");
    arg.setInt    ("type",  TYPE_U32);
    arg.setInt    ("dir",   DIR_IO);
    ft.setTable   ("arg",   arg);

    arg = FieldTable ();
    arg.setString ("name", "body");
    arg.setInt    ("type",  TYPE_LSTR);
    arg.setInt    ("dir",   DIR_IO);
    ft.setTable   ("arg",   arg);

    buf.put (ft);

   // Events
}

void Broker::writeConfig (Buffer& buf)
{
    configChanged = false;

    writeTimestamps    (buf);
    buf.putLongLong    (0);
    buf.putShort       (port);
    buf.putShort       (workerThreads);
    buf.putShort       (maxConns);
    buf.putShort       (connBacklog);
    buf.putLong        (stagingThreshold);
    buf.putShortString (storeLib);
    buf.putOctet       (asyncStore ? 1 : 0);
    buf.putShort       (mgmtPubInterval);
    buf.putLong        (initialDiskPageSize);
    buf.putLong        (initialPagesPerQueue);
    buf.putShortString (clusterName);
    buf.putShortString (version);
}

void Broker::doMethod (string  methodName,
                       Buffer& inBuf,
                       Buffer& outBuf)
{
    if (methodName.compare ("echo") == 0)
    {
        ArgsBrokerEcho args;
        uint32_t       result;

        args.io_sequence = inBuf.getLong ();
        inBuf.getLongString (args.io_body);

        result = coreObject->ManagementMethod (1, args);

        outBuf.putLong        (result);
        outBuf.putShortString ("OK");
        outBuf.putLong        (args.io_sequence);
        outBuf.putLongString  (args.io_body);
    }

    // TODO - Remove this method prior to beta
    else if (methodName.compare ("crash") == 0)
    {
        assert (0);
    }
    else
    {
        outBuf.putLong        (1);
        outBuf.putShortString ("Unknown Method");
    }
}

